Continued to overhaul client

ssmccoy opened the pull request:
Major overhaul of client including improved tests and some parts of a testing framework, removed redundant buffering and improved error handling.

Commit:
client: gearman client now faster and better

Eliminated the strange usage of in favor of using channels. Removed the
redundant `processLoop` goroutine, the `readLoop` now inlines the
processing (it was serial anyway!). Switched to the use of threadsafe
memory object pools for request and response objects, and now allows
pipelining if multiple goroutines are making requests at once.

Removed a lot of redundancy in the code, and eliminated the use of the
strange `responseHandlerMap` in favor of a concurrent map for holding
callbacks on a per-handler basis.

Refactored the `request` type such that it now uses a builder pattern,
and now itself keeps a small byte buffer for each request which is
reused when parameters are attached to requests. Streamlined the
formatting of outbound messages such that they are encoded as they are
written to the output buffer, to reduce the number of copies involved in
most operations.

Eliminated the strange blind-read loop which pretended it had no idea
how large of a buffer it would need to allocate to read a response, and
now allocates a buffer for each payload, the size of the payload as
specified by the server, respectively.

Commit:
client: Improved error handling, added snapshot test

Added a simple tool to the client test to replay snapshots captured by
`tcpflow` so long as those snapshots have a manifest indicating what the
ip and port of the sending and receiving processes were. This can be
used to capture any conversation with gearman and replay it; the tool
should probably be factored out into a common set of TCP protocol
testing utilities.

Captured a client/server conversation with a the perl `Gearman::Client`
and `Gearman::Worker` APIs against a C++ `gearmand` and verified that
the client handles the replay as expected.

Eliminated the remaining `sync.Mutex` from the client library in favor
of using compare-and-swap to access the pointer to the connection (which
is exposed as an interface value type). It turns out that interface
value types in Go are a pair of pointers represented typically as a 128
bit value, and are immutable from within the programming language. As a
simple work-around, a pointer to a struct which includes that interface
through composition was used in its stead, and CAS was used on that
pointer. This allowed the pointer to generally be used as-is without
having to be derferenced or having to deal with a pointer to an
interface (which has two levels of possibly-nil values).

Added the ability for the server stream to return an `ERROR` message as
a response to `SUBMIT_JOB_*` which in practice is what `gearmand` will
do when it was not able to succcessfully enqueue the job. This cleans up
the behavior of a client if one of the requests it makes does yield an
error.

Further, removed the additional level of buffering on `net.Conn`
connections; under the covers this is buffered non-blocking I/O anyway
so an additional `bufio` buffer was completely unnecessarry.

jacksonrnewhouse commented on client/client.go:
```diff
@@ -21,99 +22,110 @@ import (
+	expected  chan *Response
+	outbound  chan *request
+	conn      *connection
+	//rw        *bufio.ReadWriter
```
delete this?

jacksonrnewhouse commented on client/client.go:
```diff
@@ -21,99 +22,110 @@ import (
-	r.Lock()
-	r.holder[key] = rh
-	r.Unlock()
+	return
```
Is this return necessary? idiomatic?

ssmccoy replied:
It is required, because the prototype specifies return values. The values are defined as the identifiers specified in the prototype, see `* return values`; above

Commit:
coverage: fixed the coverage thinger

ssmccoy commented:
@jacksonrnewhouse The tests fail but not because they actually fail; the client tests are working the rest seem to continue to be broken.

jacksonrnewhouse commented on client/client.go:
```diff
@@ -146,29 +162,34 @@ func (client *Client) reconnect(err error) error {
+		unsafe.Pointer(&connection{conn}))
+
+	if !swapped {
+		return fmt.Errorf("client: unable to connect")
```
Why does a failure to swap mean unable to connect? Is this using CompareAndSwapPointer as an equals operator? Why not just equals?

ssmccoy replied:
So there is an equals above, before doing this whole routine (see `* equals`); in this case I think you are right we do not want to error, but if the compare-and-swap fails it means this goroutine lost the race to create the connection and the connection should be discarded.

I will change this such that

```
if !swapped {
    conn.Close()
}
```

jacksonrnewhouse replied:
Decided not to change?

jacksonrnewhouse commented on client/client.go:
```diff
@@ -218,28 +240,36 @@ func (client *Client) readLoop() {
+
+	case rt.PT_StatusRes, rt.PT_JobCreated, rt.PT_EchoRes:
+		// NOTE Anything which reads from `expected` must return the
+		// response object to the pool.
```
Why not just use defer to have this happen?

ssmccoy replied:
So; the response is given to the channel, meaning the goroutine which is yielding is expected to use it. There is _no_ guarantee that that goroutine will be _done_ with it by the time execution of this goroutine resumes. As a result; placing the element back in the pool could cause it to be mutated while being read by the other goroutine. In practice, this probably _usually wouldn't_ happen, but it _absolutely could_; so it's up to the reader of this channel to return the element to the pool.

Sometimes however, this function actually terminates processing of the `Response` struct; so the comment is just to identify the convention. Can and will move it up to the top of this `switch` statement.

jacksonrnewhouse commented on client/client.go:
```diff
@@ -180,17 +201,18 @@ func (client *Client) readLoop() {
 		}
 
-		resp = getResponse()
+		resp = client.responsePool.Get().(*Response)
```
This seems to rely on client.process() always calling client.responsePool.Put() at the end, which it seems to do but could be brittle. Could defer do the trick?

ssmccoy replied:
Actually if one of the participants fails to return something to the pool, this is resilient — the `New` function is invoked and a new object is constructed. Returning objects to the pool is just an optimization, and its failure mode is very low-cost and will not create a memory leak but simply not reuse resources.

jacksonrnewhouse replied:
The pool doesn't have a cap, I take it?

jacksonrnewhouse commented on client/client.go:
```diff
@@ -180,17 +201,18 @@ func (client *Client) readLoop() {
+
 		if err != nil {
 			client.err(err)
 			continue
```
This continue does nothing, right? I know it isn't your code, but trying to figure out golang.

ssmccoy replied:
It returns to the beginning of the for loop bypassing the below switch statement. Actually now that I'm looking at this; this code should be returning the `resp` object to the pool.

Tsk tsk...

jacksonrnewhouse replied:
oh, didn't realize it was in the for loop for some reason.

jacksonrnewhouse commented on client/client.go:
```diff
@@ -249,95 +279,63 @@ func (client *Client) err(e error) {
-	}
-	return resp
-}
+func (client *Client) submit(pt rt.PT, funcname string, arg []byte) (string, error) {
```
you could also define `var err error` in the return type, right? What's best practice there?

jacksonrnewhouse commented on client/client.go:
```diff
@@ -249,95 +279,63 @@ func (client *Client) err(e error) {
 	switch flag {
 	case rt.JobLow:
-		datatype = rt.PT_SubmitJobLow
+		pt = rt.PT_SubmitJobLow
```
can you make this packettype?

ssmccoy replied:
It's `pt` in various types and things in the code, I wanted to be more consistent. It oscollates between `pt` (short for _packet type_) and `datatype` which is mildly maddening.

It's common for Go programs to use C-style conventions; so I do like sticking with `pt` better.

jacksonrnewhouse commented on client/client.go:
```diff
@@ -249,95 +279,63 @@ func (client *Client) err(e error) {
 // Call the function and get a response.
 // flag can be set to: JobLow, JobNormal and JobHigh
-func (client *Client) Do(funcname string, data []byte,
+func (client *Client) Do(funcname string, arg []byte,
```
Is there a more descriptive name for arg, maybe `payload`?

ssmccoy replied:
I can call it `payload`. I like that better; or `contents`.

jacksonrnewhouse commented on client/client_test.go:
```diff
@@ -1,164 +1,239 @@
 
 	rt "github.com/ssmccoy/g2/pkg/runtime"
-	"github.com/appscode/go/log"
+	//"time"
```
delete commented out code

jacksonrnewhouse commented on client/request.go:
```diff
@@ -1,44 +1,40 @@
-	DataType rt.PT
-	Data     []byte
+	pt   rt.PT
+	data [][]byte
```
Is Data still in use? It is basically `[]byte`, and you replace `Data []byte` with `data [][]byte` here.

ssmccoy replied:
No I removed `Data []byte` from the type. Basically, the specification of the protocol is such that the parameters to a specific packet are terminated with `\0`; with the last one generally being the "general payload" [and as such can have nulls].

So when the request requires, for instance `function name`, `id`, `data` it will be "function\0id\0data"; well, it's easier for the callers to simply give me a slice of slices (`[][]byte`) than it to have the callers all worry about their null separators.

jacksonrnewhouse reviewed (approved):
Bunch of comments that are more like questions. Would love if you have time to instruct me in the go ways.

ssmccoy commented on client/client.go:
```diff
@@ -21,99 +22,110 @@ import (
-	holder map[string]ResponseHandler
-}
+// Return a client.
+func New(network, addr string) (client *Client, err error) {
```
`* return values`

ssmccoy commented on client/client.go:
```diff
@@ -125,6 +137,10 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
 }
 
 func (client *Client) reconnect(err error) error {
+	if client.conn != nil {
```
`* equals`

ssmccoy commented on client/client.go:
```diff
@@ -146,29 +162,34 @@ func (client *Client) reconnect(err error) error {
 
 	for client.conn != nil {
-		if _, err = io.ReadFull(client.rw, header); err != nil {
+		if _, err = io.ReadFull(*client.conn, header); err != nil {
```
This is incorrect, oops.  I mean, it works okay but we don't _need_ to dereference this pointer and such shouldn't.

Commit:
runtime: reworked stringification of packettype

Decided to go with a basic table instead of this really strange
conditional arithmetic that was being used previously.

Commit:
client: Responses to code review feedback

In response to the outstanding review of this code, moved some comments
to make their intention clearer, handled some missing conditions which
would not return response objects to the pool under certain unlikely
errors, cleaned up some of the identifier names and removed some
commented out code.

Commit:
client: Close connection when race is lost

When two callers are racing to replace the currently actively connection
and one of them loses, close the redundant connection.

Commit:
client: Remove unnecessary dereferencing

Removed the unnecessary dereferencing of the pointer to the connection
struct.
